ibtk/include/ibtk/SnapshotCache.h

```diff
@@ -0,0 +1,226 @@
+     *
+     * \note If this object is being created from a restart, the grid geometry used the patch hierarchy must be
+     * supplied. An error will be thrown if the grid geometry is an invalid pointer.
+     */
+    SnapshotCache(std::string object_name,
+                  SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                  SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db = nullptr,
+                  bool register_for_restart = false,
+                  SAMRAI::tbox::Pointer<SAMRAI::hier::GridGeometry<NDIM> > grid_geom = nullptr);
+
+    /*!
+     * \brief The destructor for class SnapshotCache deallocates patch data as needed.
+     */
+    ~SnapshotCache();
+
+    /*!
+     * Clear the snapshots stored in this object. This deallocates all patch indices stored by this object.
+     */
+    void clearSnapshots();
+
+    /*!
+     * Create a snapshot. This creates a snapshot of the current hierarchy, and copies that data from this patch index
+     * into a patch index maintained by this object. If you want to overwrite a snapshot, you should call
+     * updateSnapshot() instead. Data is not allocated in this class until this function is called.
+     *
+     */
+    void
+    setSnapshot(int u_idx, double time, SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > current_hierarchy);
+
+    /*!
+     * Update a snapshot. The data from this patch index is copied into an array of indices controlled by this
+     * object. This outputs an error if a snapshot with that time point within the specified tolerance can not be found.
+     *
+     * Calls to this function may use a different patch index than the one used in setSnapshot(), but the underlying
+     * data layout must be consistent. This data layout is specified by the variable given to the object in the
+     * constructor.
+     */
+    void updateSnapshot(int u_idx,
+                        double time,
+                        SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > current_hierarchy,
+                        double tol = 1.0e-8);
+
+    /*!
+     * Transfer the snapshot at the specified time value within a given tolerance to the patch index u_idx on the
+     * supplied patch hierarchy. The patch index u_idx must contain sufficient ghost cell width to perform the
+     * operations used by the refinement operator.
+     *
+     * Note this can cause a significant amount of communication if the supplied patch hierarchy has a different
+     * configuration of patches than the snapshot patch hierarchy.
+     *
+     * This function does not synchronize the data on the current hierarchy (i.e. no coarsening is performed). It
+     * transfers the snapshot using the refinement operator.
+     */
+    //@{
+    void getSnapshot(int u_idx,
+                     double time,
+                     SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > current_hierarchy,
```

Comment:
These functions should just return a `tbox::Pointer<hier::PatchHierarchy>` by value.

Reply:
The hierarchy is not a return here. We interpolate the snapshot from the snapshotted hierarchy onto `current_hierarchy` and store it in `u_idx`.